File: slidestation/stats.py
"""Progress across the whole library: slides per hour, trays left, when the target is reached.

Pure functions over session.json dicts, mirrored in frontend/src/lib/stats.ts (the browser
version) - keep the two identical. A slide's work is timed by `developed_at` (set when it is marked
developed) or, for slides uploaded without being developed first, by its upload time
(`immich["at"]`). Trays from before either existed simply don't contribute to the rate.
"""
from __future__ import annotations

import time
from datetime import datetime

DEFAULT_TARGET = 10_000
BREAK_S = 10 * 60  # a longer gap between two slides is a break, not work
PACE_DAYS = 14  # the projected finish uses the pace of the last two weeks
MIN_ACTIVE_S = 5 * 60  # less work than this says nothing about a rate yet


def slide_times(d: dict) -> list[float]:
    """When each slide of a tray was finished (developed, else uploaded), where that's known."""
    out = []
    for g in d.get("groups", []):
        t = g.get("developed_at") or (g.get("immich") or {}).get("at")
        if t:
            out.append(float(t))
    return out


def per_hour(times: list[float]) -> tuple[float | None, float]:
    """Slides per hour of actual work, and the hours worked. Gaps longer than BREAK_S are breaks."""
    ts = sorted(times)
    active = sum(min(b - a, BREAK_S) for a, b in zip(ts, ts[1:]))
    if len(ts) < 2 or active < MIN_ACTIVE_S:
        return None, active / 3600
    return (len(ts) - 1) / (active / 3600), active / 3600


def library_stats(summaries: list[dict], times: list[float], target: int = DEFAULT_TARGET,
                  now: float | None = None) -> dict:
    """summaries: store.summary() of every tray; times: slide_times() of all of them together."""
    now = time.time() if now is None else now
    slides = sum(s["slides"] for s in summaries)
    done = sum(s["reviewed"] for s in summaries)  # developed, uploaded or skipped
    finished = [s for s in summaries if s["slides"] and s["uploaded"] + s["skipped"] >= s["slides"]]
    rate, hours = per_hour(times)
    recent = [t for t in times if now - t < PACE_DAYS * 86400]
    # slides a day over the days worked so far, up to the last PACE_DAYS (a first day counts as one)
    per_day = len(recent) / min(PACE_DAYS, max(1.0, (now - min(recent)) / 86400)) if recent else None
    remaining = max(0, target - done)
    finish = None
    if remaining == 0:
        finish = datetime.fromtimestamp(now).strftime("%Y-%m-%d")
    elif per_day:
        finish = datetime.fromtimestamp(now + remaining / per_day * 86400).strftime("%Y-%m-%d")
    per_tray = slides / len(summaries) if summaries and slides else None
    midnight = datetime.fromtimestamp(now).replace(hour=0, minute=0, second=0, microsecond=0).timestamp()
    return {
        "target": target,
        "trays": {"total": len(summaries), "finished": len(finished), "open": len(summaries) - len(finished)},
        "slides": {
            "total": slides,
            "done": done,
            "uploaded": sum(s["uploaded"] for s in summaries),
            "skipped": sum(s["skipped"] for s in summaries),
            "to_develop": slides - done,
        },
        "remaining": remaining,
        "per_hour": round(rate, 1) if rate else None,
        "hours_worked": round(hours, 2),
        "hours_left": round(remaining / rate, 1) if rate else None,
        "per_day": round(per_day, 1) if per_day else None,
        "today": sum(1 for t in times if t >= midnight),
        "last_7_days": sum(1 for t in times if now - t < 7 * 86400),
        "finish": finish,
        # trays still to scan, at the slides per tray so far
        "trays_to_scan": max(0, round((target - slides) / per_tray)) if per_tray else None,
    }

File: frontend/src/lib/stats.ts
// Progress across the whole library (slidestation/stats.py, for the browser version) — keep the two
// identical. A slide's work is timed by `developed_at`, or for slides uploaded without being
// developed first, by its upload time (`immich.at`).
import type { Summary } from "@/lib/api";

export const DEFAULT_TARGET = 10_000;
export const BREAK_S = 10 * 60; // a longer gap between two slides is a break, not work
export const PACE_DAYS = 14; // the projected finish uses the pace of the last two weeks
const MIN_ACTIVE_S = 5 * 60; // less work than this says nothing about a rate yet

export type Stats = {
  target: number;
  trays: { total: number; finished: number; open: number };
  slides: { total: number; done: number; uploaded: number; skipped: number; to_develop: number };
  remaining: number;
  per_hour: number | null;
  hours_worked: number;
  hours_left: number | null;
  per_day: number | null;
  today: number;
  last_7_days: number;
  /** YYYY-MM-DD, local time. */
  finish: string | null;
  trays_to_scan: number | null;
};

type TimedGroup = { developed_at?: number; immich?: { at?: number } | null };

/** When each slide of a tray was finished (developed, else uploaded), where that's known. */
export const slideTimes = (d: { groups: TimedGroup[] }): number[] =>
  d.groups.flatMap((g) => {
    const t = g.developed_at || g.immich?.at;
    return t ? [t] : [];
  });

/** Slides per hour of actual work, and the hours worked. Gaps longer than BREAK_S are breaks. */
export function perHour(times: number[]): [number | null, number] {
  const ts = [...times].sort((a, b) => a - b);
  let active = 0;
  for (let i = 1; i < ts.length; i++) active += Math.min(ts[i] - ts[i - 1], BREAK_S);
  if (ts.length < 2 || active < MIN_ACTIVE_S) return [null, active / 3600];
  return [(ts.length - 1) / (active / 3600), active / 3600];
}

const round = (x: number, digits: number) => Math.round(x * 10 ** digits) / 10 ** digits;
const ymd = (t: Date) =>
  `${t.getFullYear()}-${String(t.getMonth() + 1).padStart(2, "0")}-${String(t.getDate()).padStart(2, "0")}`;

type Counts = Pick<Summary, "slides" | "reviewed" | "uploaded" | "skipped">;

/** summaries: every tray's summary; times: slideTimes() of all of them together (seconds). */
export function libraryStats(
  summaries: Counts[],
  times: number[],
  target = DEFAULT_TARGET,
  now = Date.now() / 1000,
): Stats {
  const slides = summaries.reduce((n, s) => n + s.slides, 0);
  const done = summaries.reduce((n, s) => n + s.reviewed, 0); // developed, uploaded or skipped
  const finished = summaries.filter((s) => s.slides && s.uploaded + s.skipped >= s.slides).length;
  const [rate, hours] = perHour(times);
  const recent = times.filter((t) => now - t < PACE_DAYS * 86400);
  // slides a day over the days worked so far, up to the last PACE_DAYS (a first day counts as one)
  const perDay = recent.length
    ? recent.length / Math.min(PACE_DAYS, Math.max(1, (now - Math.min(...recent)) / 86400))
    : null;
  const remaining = Math.max(0, target - done);
  const finish =
    remaining === 0
      ? ymd(new Date(now * 1000))
      : perDay
        ? ymd(new Date((now + (remaining / perDay) * 86400) * 1000))
        : null;
  const perTray = summaries.length && slides ? slides / summaries.length : null;
  const midnight = new Date(now * 1000);
  midnight.setHours(0, 0, 0, 0);
  return {
    target,
    trays: { total: summaries.length, finished, open: summaries.length - finished },
    slides: {
      total: slides,
      done,
      uploaded: summaries.reduce((n, s) => n + s.uploaded, 0),
      skipped: summaries.reduce((n, s) => n + s.skipped, 0),
      to_develop: slides - done,
    },
    remaining,
    per_hour: rate ? round(rate, 1) : null,
    hours_worked: round(hours, 2),
    hours_left: rate ? round(remaining / rate, 1) : null,
    per_day: perDay ? round(perDay, 1) : null,
    today: times.filter((t) => t >= midnight.getTime() / 1000).length,
    last_7_days: times.filter((t) => now - t < 7 * 86400).length,
    finish,
    // trays still to scan, at the slides per tray so far
    trays_to_scan: perTray ? Math.max(0, Math.round((target - slides) / perTray)) : null,
  };
}
